### PR TITLE
Deprecate jsonformoderncpp in favor of nlohmann_json

### LINKED ISSUE #307: [package] jsonformoderncpp/all: cmake exported name is different than upstream
according to upstream:
https://github.com/nlohmann/json#cmake

cmake should use `nlohmann_json` as target name.
conan index is using `jsonformoderncpp` and `jsonformoderncpp::jsonformoderncpp`

this should be fixed with e.g. `self.cpp_info.name` in the recipe

### PR DESCRIPTION
**Pull request by IceflowRE**
Fix #307

Specify library name and version:  **lib/1.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
I think is related to Cmake target, not the entire recipe name.

Did you try using `self.cpp_info.name = "nlohmann_json"` ?

**Comment by IceflowRE** [comment]
I thought its a bit weird if you use another recipe name than the cmake package in code at the end. I mean its a bit weird at all that you name a project jsonformoderncpp but the package name is different.

**Comment by Croydon** [comment]
I agree, that the upstream behaviour seems confusing

Is jsonformonderncpp packaged for other package managers we could look at as a reference?
Or maybe we should simply ask upstream what to do with this name mismatch

Most of the time, there might be a difference in upper/lowercase, dashes or added/missing version numbers, but in this case it is a completely different name 🤔

**Comment by gocarlos** [comment]
- https://github.com/openembedded/meta-openembedded/tree/master/meta-oe/recipes-devtools/nlohmann-json
- https://github.com/microsoft/vcpkg/tree/master/ports/nlohmann-json
- https://packages.debian.org/stretch/nlohmann-json-dev

**Comment by SSE4** [comment]
@lasote @danimtb seems like we have an issue with such kind of rename PRs

**Comment by danimtb** [comment]
Seems the build service is still not prepared for this kind of renamings. Additionally, we don't have a policy set about how to proceed in this kind of cases. In order to move this forward, I'd suggest instead of renaming, creating a new folder `nlohmann-json` with the recipe and the required changes.

**Comment by uilianries** [comment]
what about an alias?

**Comment by danimtb** [comment]
As we are still discovering the possibilities under the EAP, I feel we should avoid creating alias as we are not braking anyone right now. Let's change the folder and we will discuss how to deal with the *old* recipe name

**Comment by Croydon** [comment]
Personally, I feel like we do NOT break anyone by removing recipes and/or version from this git repository.

We would break workflows when we would deleted uploads from Conan Center, but we never do this.

So in my opinion, it is totally fine when we remove recipes and/or recipe versions from this repository

**Comment by theodelrieu** [comment]
Are we ready to rename? So that we can merge #312?

**Comment by danimtb** [comment]
Please follow my comments in #307

**Comment by uilianries** [comment]
@IceflowRE https://github.com/IceflowRE/conan-center-index/pull/1

File: recipes/jsonformoderncpp/all/conanfile.py [hunk 1]
```diff
@@ -26,6 +26,9 @@ def source(self):
         tools.get(**self.conan_data["sources"][self.version])
         extracted_dir = "json-" + self.version
         os.rename(extracted_dir, self._source_subfolder)
+
```
Comment (danimtb) [inline recipes/jsonformoderncpp/all/conanfile.py]:
```suggestion
```

File: recipes/jsonformoderncpp/all/conanfile.py [hunk 2]
```diff
@@ -26,6 +26,9 @@ def source(self):
         tools.get(**self.conan_data["sources"][self.version])
         extracted_dir = "json-" + self.version
         os.rename(extracted_dir, self._source_subfolder)
+        
+    def configure(self):
```
Comment (danimtb) [inline recipes/jsonformoderncpp/all/conanfile.py]:
```suggestion
```

File: recipes/jsonformoderncpp/all/conanfile.py [hunk 3]
```diff
@@ -26,6 +26,9 @@ def source(self):
...
+        self.output.warn("[DEPRECATED] Package jsonformoderncpp is being deprecated. Change yours to require nlohmann_json instead")
```
Comment (danimtb) [inline recipes/jsonformoderncpp/all/conanfile.py]:
```suggestion
```

File: recipes/nlohmann_json/config.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
```
Comment (danimtb) [inline recipes/nlohmann_json/config.yml]:
```suggestion
```

File: recipes/nlohmann_json/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        self.copy(pattern="LICENSE*", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib"))
+        try:
+            os.remove(os.path.join(self.package_folder, "nlohmann_json.natvis"))
+        except FileNotFoundError:
+            pass
+
+
```
Comment (danimtb) [inline recipes/nlohmann_json/all/conanfile.py]:
```suggestion
```

File: recipes/nlohmann_json/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    def package_id(self):
+        self.info.header_only()
```
Comment (danimtb) [inline recipes/nlohmann_json/all/conanfile.py]:
```suggestion
        self.info.header_only()

```

**Comment by IceflowRE** [comment]
I removed the other lines in the other PR, trying to avoid a change in "another" recipe.

**Pull request merged**